### PR TITLE
Support configuring php-fpm pm via environment variables

### PR DESCRIPTION
**Pull request by renatomefi**
# Usabilla PHP Docker Template

Reviewers: @usabilla/oss-docker

First pass for comment of https://github.com/usabilla/php-docker-template/issues/119

## Type

Please specify the type of changes being proposed:

| Q                 | A
| ----------------- | -----
| Documentation?    | TODO
| Dockerfile change?| yes
| Build feature?    | no
| Apply CVE Patch?  | no
| Remove CVE Patch? | no

## Changelog

- Please make sure you have checked our [Contributing guidelines](https://github.com/usabilla/php-docker-template/blob/master/.github/CONTRIBUTING.md)

### CODE REVIEW

File: test/container/php/test_fpm_entrypoint.py [hunk 1]
```diff
@@ -11,3 +11,56 @@ def test_php_fpm_exec_has_dumb_init(host):
...
+    docker-php-entrypoint-init php-fpm
+    ''')
+
+    # We already have a running fpm, thus it's fine to have an error
+    # since the objective is to test the templating
+    assert replace_template.rc is 78
+    assert "ERROR: Another FPM instance seems to already listen on /var/run/php-fpm.sock" in replace_template.stderr
+    assert "ERROR: FPM initialization failed" in replace_template.stderr
+
+    config = config_file = host.file("/usr/local/etc/php-fpm.d/zz-docker.conf")
```
Comment (bendavies) [inline test/container/php/test_fpm_entrypoint.py]: you can also run `php-fpm -tt` to assert that fpm is actually using the config in this file.

Comment (renatomefi) [reply test/container/php/test_fpm_entrypoint.py]: Good idea, to check the effective configuration, will add that! Thanks

File: test/container/php/test_fpm_entrypoint.py [hunk 2]
```diff
@@ -11,3 +11,67 @@ def test_php_fpm_exec_has_dumb_init(host):
...
+
+    replace_template = host.run('''
+    ENV PHP_FPM_PM=static \
+    ENV PHP_FPM_PM_MAX_CHILDREN=70 \
+    ENV PHP_FPM_PM_START_SERVERS=10 \
+    ENV PHP_FPM_PM_MIN_SPARE_SERVERS=20 \
+    ENV PHP_FPM_PM_MAX_SPARE_SERVERS=40 \ 
+    ENV PHP_FPM_PM_PROCESS_IDLE_TIMEOUT=35 \
+    ENV PHP_FPM_PM_MAX_REQUESTS=500 \
+    docker-php-entrypoint-init php-fpm
```
Comment (bendavies) [inline test/container/php/test_fpm_entrypoint.py]: can maybe do `-tt` here then we won't have the `already running` message below?

Comment (renatomefi) [reply test/container/php/test_fpm_entrypoint.py]: I'll give that a try!

Comment (renatomefi) [reply test/container/php/test_fpm_entrypoint.py]: Done, check it out!

File: test/container/php/test_fpm_entrypoint.py [hunk 3]
```diff
@@ -11,3 +11,65 @@ def test_php_fpm_exec_has_dumb_init(host):
...
+    PHP_FPM_PM_START_SERVERS=10
+    PHP_FPM_PM_MIN_SPARE_SERVERS=20
+    PHP_FPM_PM_MAX_SPARE_SERVERS=40 
+    PHP_FPM_PM_PROCESS_IDLE_TIMEOUT=35
+    PHP_FPM_PM_MAX_REQUESTS=500
+    docker-php-entrypoint-init php-fpm -tt
+    ''')
+
+    assert replace_template.rc is 0
+    assert "NOTICE: configuration file /usr/local/etc/php-fpm.conf test is successful" in replace_template.stderr
```
Comment (bendavies) [inline test/container/php/test_fpm_entrypoint.py]:
```
    assert "pm = static" in replace_template.stderr
    assert "pm.max_children = 70" in replace_template.stderr
    assert "pm.start_servers = 10" in replace_template.stderr
    assert "pm.min_spare_servers = 20" in config
    assert "pm.max_spare_servers = 40" in replace_template.stderr
    assert "pm.process_idle_timeout = 35" in replace_template.stderr
    assert "pm.max_requests = 500" in replace_template.stderr
```
 as well?

Comment (renatomefi) [reply test/container/php/test_fpm_entrypoint.py]: In this case it won't work, `php-fpm -tt` will take the effective configuration, for that we'd have to restart fpm and that's the part I didn't do!

Comment (bendavies) [reply test/container/php/test_fpm_entrypoint.py]: aha, very good.

Comment (renatomefi) [reply test/container/php/test_fpm_entrypoint.py]: Sadly php-fpm doesn't have a nice reload mechanism, meaning I'd have to find a way to bring it down and up without killing the container, sounded too complex to be worth it!

**Comment by renatomefi** [comment]
Thanks so much @bendavies for the nice work! Let me know if you're still willing to open a PR for the docs!

**Comment by bendavies** [comment]
yes, i can do it when i get a moment.

File: src/php/fpm/conf/zz-docker.conf.template [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+pm.min_spare_servers = ${PHP_FPM_PM_MIN_SPARE_SERVERS}
+
+; The desired maximum number of idle server processes.
+; Note: Used only when pm is set to 'dynamic'
+; Note: Mandatory when pm is set to 'dynamic'
+pm.max_spare_servers = ${PHP_FPM_PM_MAX_SPARE_SERVERS}
+
+; The number of seconds after which an idle process will be killed.
+; Note: Used only when pm is set to 'ondemand'
+; Default Value: 10s
```
Comment (WyriHaximus) [inline src/php/fpm/conf/zz-docker.conf.template]: This was already mention here https://github.com/usabilla/php-docker-template/pull/124#discussion_r327188761 but maybe this makes it a bit less error prone:

```suggestion
; Default Value (seconds): 10
```

Comment (renatomefi) [reply src/php/fpm/conf/zz-docker.conf.template]: This is how the official file is :/

**Pull request merged**